tools/c7n_azure/c7n_azure/policy.py

```diff
@@ -147,18 +159,127 @@ def get_logs(self, start, end):
         raise NotImplementedError("error - not implemented")
 
 
-@execution.register('azure-stream')
-class AzureStreamMode(AzureFunctionMode):
+@execution.register(CONST_AZURE_EVENT_TRIGGER_MODE)
+class AzureEventMode(AzureFunctionMode):
     """A policy that runs/executes in azure functions from an
-    azure activity log stream."""
+    azure event."""
+
+    schema = utils.type_schema(CONST_AZURE_EVENT_TRIGGER_MODE,
+                               events={'type': 'array', 'items': {
+                                   'oneOf': [
+                                       {'type': 'string'},
+                                       {'type': 'object',
+                                        'required': ['resourceProvider', 'event'],
+                                        'properties': {
+                                            'resourceProvider': {'type': 'string'},
+                                            'event': {'type': 'string'}}}]
+                               }},
+                               required=['events'],
+                               rinherit=AzureFunctionMode.schema)
+
+    def provision(self):
+        super(AzureEventMode, self).provision()
+        key = self.get_webhook_key()
+        webhook_url = 'https://%s.azurewebsites.net/api/%s?code=%s' % (self.webapp_name,
+                                                                       self.policy_name, key)
+        destination = WebHookEventSubscriptionDestination(
+            endpoint_url=webhook_url
+        )
+        event_filter = EventSubscriptionFilter()
+        event_info = EventSubscription(destination=destination, filter=event_filter)
+        scope = '/subscriptions/%s' % self.session.subscription_id
+
+        #: :type: azure.mgmt.eventgrid.EventGridManagementClient
+        eventgrid_client = self.session.client('azure.mgmt.eventgrid.EventGridManagementClient')
+
+        status_success = False
+        while not status_success:
+            try:
+                event_susbcription = eventgrid_client.event_subscriptions.create_or_update(
+                    scope, self.webapp_name, event_info)
+
+                event_susbcription.result()
+                self.log.info('Event subscription creation succeeded')
+                status_success = True
+            except CloudError as e:
+                self.log.info(e)
+                self.log.info('Retrying in 30 seconds')
+                time.sleep(30)
+
+    def get_webhook_key(self):
+        token_headers = {
+            'Authorization': 'Bearer %s' % self.session.get_bearer_token()
+        }
 
-    schema = utils.type_schema('azure-stream', rinherit=AzureFunctionMode.schema)
+        key_url = (
+            'https://management.azure.com'
```

Comment:
Worth moving this URL template to a constant, because we have api-version hard coded

Reply:
We only use it one place, but I'll refactor it out. Typically I think constants are more useful in scenarios where you will reuse it. Also - this one is tricky because this isn't a documented api, so I'm not sure if this will even be the appropriate format going forward if they decide to change it.